test/view/generate_view.cpp

```diff
@@ -0,0 +1,132 @@
+#include <stl2/view/generate_view.hpp>
+
+#include "../simple_test.hpp"
+#include <cstring>
+#include <stl2/ranges.hpp>
+#include <stl2/view/take_exactly.hpp>
+
+namespace ranges = __stl2;
+
+namespace view {
+	using namespace ranges::view;
+	using namespace ext;
+} // namespace view
+
+// a simple type to test move semantics
+struct MoveOnlyString {
+	char const* sz_;
+
+	MoveOnlyString(char const* sz = "")
+	  : sz_(sz)
+	{}
+	MoveOnlyString(MoveOnlyString&& that)
+	  : sz_(that.sz_)
+	{
+		that.sz_ = "";
+	}
+	MoveOnlyString(MoveOnlyString const&) = delete;
+	MoveOnlyString& operator=(MoveOnlyString&& that)
+	{
+		sz_ = that.sz_;
+		that.sz_ = "";
+		return *this;
+	}
+	MoveOnlyString& operator=(MoveOnlyString const&) = delete;
+	bool operator==(MoveOnlyString const& that) const
+	{  return 0 == std::strcmp(sz_, that.sz_); }
+
+	bool operator<(const MoveOnlyString& that) const
+	{ return std::strcmp(sz_, that.sz_) < 0; }
+
+	bool operator!=(MoveOnlyString const& that) const
+	{ return !(*this == that); }
+
+	friend std::ostream& operator<<(std::ostream& sout, MoveOnlyString const& str)
+	{ return sout << '"' << str.sz_ << '"'; }
+};
+
+struct MoveOnlyFunction {
+	MoveOnlyString str_;
+	int i_;
+
+	char operator()()
+	{ return str_.sz_[i_++]; }
```

Comment:
I'm concerned that this doesn't explode *very loudly* if asked to access past-the-end of the string.

Reply:
I hope `STL2_EXPECT` is loud enough (I can also make this change upstream if you like... there's a `RANGES_EXPECT`, right?).

Reply:
> I hope `STL2_EXPECT` is loud enough 

Yes, it is.

>(I can also make this change upstream if you like... there's a `RANGES_EXPECT`, right?).

That would be nice. And yes there is, and I strongly suspect that they have the same behavior. ;)